### PR TITLE
Integrating two-up

### PR DESCRIPTION
**Pull request by jakearchibald**
Ready for review!

### CODE REVIEW

**Comment by jakearchibald** [comment]
Pixelated bugs:

Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=843615.
Safari: https://bugs.webkit.org/show_bug.cgi?id=185683.

I need to add `image-rendering: optimizespeed;` for Firefox.
I need to fix my test to work in Edge 😢.

File: src/components/output/index.tsx [hunk 1]
```diff
@@ -10,34 +14,88 @@ type State = {};
...
-
-  updateCanvas(img: ImageBitmap) {
-    if (!this.canvas) return;
-    const ctx = this.canvas.getContext('2d');
-    if (!ctx) return;
-    ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
-    ctx.drawImage(img, 0, 0);
+  updateCanvases(img: ImageBitmap) {
+    for (const [i, canvas] of [this.canvasLeft, this.canvasRight].entries()) {
+      if (!canvas) return;
```
Comment (surma) [inline src/components/output/index.tsx]: `continue`? (If there is `canvasRight`, but not `canvasLeft`?)

File: src/components/output/index.tsx [hunk 2]
```diff
@@ -10,34 +14,88 @@ type State = {};
 
 export default class App extends Component<Props, State> {
   state: State = {};
-  canvas?: HTMLCanvasElement;
+  canvasLeft?: HTMLCanvasElement;
+  canvasRight?: HTMLCanvasElement;
+  pinchZoomLeft?: PinchZoom;
```
Comment (surma) [inline src/components/output/index.tsx]: Considering we render those in the JSX, can they ever be null?

File: src/components/output/index.tsx [hunk 3]
```diff
@@ -10,34 +14,88 @@ type State = {};
...
+    if (!this.pinchZoomLeft) return;
+    // If the event is on the handle of the two-up, let it through.
+    if (targetEl.closest('.' + twoUpHandle)) return;
+    // If we've already retargeted this event, let it through.
+    if (this.retargetedEvents.has(event)) return;
+    // Stop the event in its tracks.
+    event.stopImmediatePropagation();
+    event.preventDefault();
+    // Clone the event & dispatch
+    const clonedEvent = new (event.constructor as typeof Event)(event.type, event);
```
Comment (surma) [inline src/components/output/index.tsx]: Surprised you need to force the type of `event.constructor`, but I’m guessing you did try without.

File: src/components/output/index.tsx [hunk 4]
```diff
@@ -10,34 +14,88 @@ type State = {};
...
+        <two-up
+          // Event redirecting. See onRetargetableEvent.
+          onTouchStartCapture={this.onRetargetableEvent}
+          onTouchEndCapture={this.onRetargetableEvent}
+          onTouchMoveCapture={this.onRetargetableEvent}
+          onPointerDownCapture={this.onRetargetableEvent}
+          onMouseDownCapture={this.onRetargetableEvent}
+          onWheelCapture={this.onRetargetableEvent}
+        >
+          <pinch-zoom onChange={this.onPinchZoomLeftChange} ref={p => this.pinchZoomLeft = p as PinchZoom}>
```
Comment (surma) [inline src/components/output/index.tsx]: Can’t you do `p: PinchZoom => this.pinchZoomLeft = p`?

File: src/components/output/index.tsx (hunk 2 above)
Comment (jakearchibald) [reply src/components/output/index.tsx]: They're nullable in terms of TypeScript, since they're defined *after* the constructor.

File: src/components/output/index.tsx (hunk 3 above)
Comment (jakearchibald) [reply src/components/output/index.tsx]: I should add a comment linking to https://github.com/Microsoft/TypeScript/issues/3841.

File: src/components/output/index.tsx (hunk 4 above)
Comment (jakearchibald) [reply src/components/output/index.tsx]: Nah, since ref provides an `Element`. Typing the arg is different to casting, and it's casting that I need here.

Casting seems fair here, since it won't ever be something other than a canvas (famous last words).

File: src/components/output/index.tsx (hunk 1 above)
Comment (jakearchibald) [reply src/components/output/index.tsx]: I guess this is one of those "this should never happen" cases.

Should we throw errors in these cases?

Comment (surma) [reply src/components/output/index.tsx]: If it’s never supposed to happen, I’d say throw. Silent return can be frustrating

Comment (jakearchibald) [reply src/components/output/index.tsx]: Agreed. I'll fix other places where I've done this.

**Comment by developit** [comment]
_(aside: ugh, I made some comments here but never submitted the review)_

**Comment by jakearchibald** [comment]
@developit will it not let you submit it now?
